src/web/app/pages-instructor/instructor-course-edit-page/instructor-course-edit-page.component.ts

```diff
@@ -12,13 +71,627 @@ import { ActivatedRoute } from '@angular/router';
+        }, (resp: ErrorMessageOutput) => {
+          this.statusMessageService.showErrorMessage(resp.error.message);
+        });
+  }
+
+  /**
+   * Updates the stored instructor and instructor list entities.
+   */
+  updateInstructorDetails(index: number, instr: InstructorAttributes): void {
+    const newPrivileges: Privileges = instr.privileges;
+
+    // Update the stored instructor
+    if (this.instructorList.length === 1) {
+      // If there is only one instructor, the instructor can modify instructors by default
+      newPrivileges.courseLevel.canmodifyinstructor = true;
+      this.instructor = instr;
+    }
+
+    // Update elements for privileges if needed
+    if (this.instructor.googleId === instr.googleId) {
+      this.updateElementsForPrivileges();
+    }
+
+    // Update the stored instructor list
+    this.instructorList[index] = instr;
+  }
+
+  /**
+   * Gets the privileges for a particular role.
+   */
+  private getPrivilegesForRole(role: string): Privileges {
+    if (role === 'Co-owner') {
+      return this.instructorPrivileges.coowner;
+    }
+
+    if (role === 'Manager') {
+      return this.instructorPrivileges.manager;
+    }
+
+    if (role === 'Observer') {
+      return this.instructorPrivileges.observer;
+    }
+
+    if (role === 'Tutor') {
+      return this.instructorPrivileges.tutor;
+    }
+
+    return this.instructorPrivileges.custom;
+  }
+
+  /**
+   * Updates elements and buttons related to the current instructor's privileges.
+   */
+  private updateElementsForPrivileges(): void {
+    const courseBtns: HTMLCollectionOf<Element> = document.getElementsByClassName('btn-course');
+    for (const courseBtn of courseBtns as any) {
+      (courseBtn as HTMLInputElement).disabled = !this.instructor.privileges.courseLevel.canmodifycourse;
+    }
+
+    const instrBtns: HTMLCollectionOf<Element> = document.getElementsByClassName('btn-instr');
+    for (const instrBtn of instrBtns as any) {
+      (instrBtn as HTMLInputElement).disabled = !this.instructor.privileges.courseLevel.canmodifyinstructor;
+    }
+  }
+
+  /**
+   * Toggles the add instructor form.
+   */
+  toggleIsAddingInstructor(): void {
+    this.isAddingInstructor = !this.isAddingInstructor;
+
+    if (this.isAddingInstructor) {
+      this.initAddInstructorForm();
+    }
+  }
+
+  /**
+   * Initialises a new form for adding an instructor to the current course.
+   */
+  private initAddInstructorForm(): void {
+    this.formAddInstructor = this.fb.group({
+      name: [''],
+      email: [''],
+      isDisplayedToStudents: [{ value: true }],
+      displayedName: ['Instructor'],
+      role: ['Co-owner'],
     });
   }
 
+  /**
+   * Adds a new instructor to the current course.
+   */
+  onSubmitAddInstructor(formAddInstructor: FormGroup): void {
+    // Create a copy of the added instructor
+    const addedInstructor: InstructorAttributes = {
+      googleId: formAddInstructor.controls.googleId.value,
```

Comment:
This is causing bug that prevents instructor from being added. When adding a new instructor, they would not have a google ID yet.